sklearn/cluster/tests/test_k_means.py

```diff
@@ -771,4 +779,55 @@ def test_x_squared_norms_init_centroids():
 def test_max_iter_error():
 
     km = KMeans(max_iter=-1)
-    assert_raise_message(ValueError, 'Number of iterations should be', km.fit, X)
+    assert_raise_message(ValueError, 'Number of iterations should be',
+                         km.fit, X)
+
+
+def test_float_precision():
+    km = KMeans(n_init=1, random_state=30)
+    mb_km = MiniBatchKMeans(n_init=1, random_state=30)
+
+    inertia = {}
+    X_new = {}
+    centers = {}
+
+    for estimator in [km, mb_km]:
+        for is_sparse in [False, True]:
+            for dtype in [np.float64, np.float32]:
+                if is_sparse:
+                    X_test = sp.csr_matrix(X_csr, dtype=dtype)
+                else:
+                    X_test = dtype(X)
+                estimator.fit(X_test)
+                # dtype of cluster centers has to be the dtype of the input data
+                assert_equal(estimator.cluster_centers_.dtype, dtype)
+                inertia[dtype] = estimator.inertia_
+                X_new[dtype] = estimator.transform(X_test)
+                centers[dtype] = estimator.cluster_centers_
+                # make sure predictions correspond to the correct label
+                assert_equal(estimator.predict(X_test[0]), estimator.labels_[0])
+                if estimator == mb_km:
+                    estimator.partial_fit(X_test[0:3])
+                    # dtype of cluster centers has to stay the same after partial_fit
+                    assert_equal(estimator.cluster_centers_.dtype, dtype)
+
+            # compare arrays with low precision since the difference between
+            # 32 and 64 bit sometimes makes a difference up to the 4th decimal place
+            assert_array_almost_equal(inertia[np.float32], inertia[np.float64],
+                                    decimal=4)
+            assert_array_almost_equal(X_new[np.float32], X_new[np.float64], decimal=4)
+            assert_array_almost_equal(centers[np.float32], centers[np.float64], decimal=4)
+
+
+def test_KMeans_init_centers():
+    # This test is used to check KMeans won't mutate the user provided input array silently
+    # even if input data and init centers have the same type
+    X_small = np.array([[1.1, 1.1], [-7.5, -7.5], [-1.1, -1.1], [7.5, 7.5]])
+    init_centers = np.array([[0.0, 0.0], [5.0, 5.0], [-5.0, -5.0]])
+    for dtype in [np.int32, np.int64, np.float32, np.float64]:
+        X_test = dtype(X_small)
+        init_centers_test = dtype(init_centers)
+        assert_equal(X_test.dtype, init_centers_test.dtype)
+        km = KMeans(init=init_centers_test, n_clusters=3)
+        km.fit(X_test)
+        assert_equal(False, km.cluster_centers_ is init_centers)
```

Comment:
I've tested it for `int` dtype [here](https://github.com/scikit-learn/scikit-learn/pull/6846/files#diff-bb423712034ea614cbea754b59ce5e31R667) and for `float` dtype [here](https://github.com/scikit-learn/scikit-learn/pull/6846/files#diff-bb423712034ea614cbea754b59ce5e31R798).